### ISSUE TITLE
samtools depth does not handle a bed file

### ISSUE THREAD

**Issue by Neato-Nick** [samtools/samtools#863]
#### Are you using the latest version of [samtools](https://github.com/samtools/samtools/releases/latest) and [HTSlib](https://github.com/samtools/htslib/releases/latest)? If not, please specify.
(run `samtools --version`)

Yes - using latest full release, Samtools 1.8. I first noticed the bug on Samtools 1.6, but upgrading to 1.8 did not resolve the issue

 #### Please describe your environment.
- OS (run `uname -sr` on Linux/Mac OS or ` wmic os get Caption, Version` on Windows)
- machine architecture (run `uname -m` on Linux/Mac OS or ` wmic os get OSArchitecture` on Windows)
- compiler (run `gcc --version` or `clang --version`)

Linux 3.10.0-327.el7.x86_64 on a high-performance computing cluster

#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.

I can not get samtools depth to recognize a bed file for where to count depth positions. The command I'm using is

`samtools depth some.bam -b some.bed`

I've tried putting the bed file in quotes (` "some.bed" `), and various ways of formatting the bed file (added/removed header, added/removed comment lines). But every time, I get the error:

```
[bed_read] Parse error reading some.bed at line 1
samtools depth: Could not read file "some.bed": No such file or directory
```

It seems like this error would be on my side, but I can't figure out what's wrong. I'm able to supply regions like `-r "Chr1:456-789"`. I'm just having issues with using a bed file.

**Comment by valeriuo** [comment]
Looks like a problem with the BED file content. Can you attach it to issue report?

**Comment by tseemann** [comment]
@Neato-Nick can you try putting the command line options first:
```
samtools depth -b some.bed some.bam
```

**Comment by Neato-Nick** [comment]
Also to try and make it work:

- switched the order of parameters like @tseemann suggested
- removed the fourth column of my BED so it was the bare minimum.

Neither worked. My BED is only a few lines, so here it is:

```
Supercontig_1.1 6645440 6645750 NW_003303758.1
Supercontig_1.18        1350077 1349821 NW_003303741.1
Supercontig_1.17        2262454 2262863 NW_003303742.1
Supercontig_1.19        1535755 1536103 NW_003303740.1
Supercontig_1.19        1535470 1535718 NW_003303740.1
Supercontig_1.489       41847   41562   NW_003303276.1
Supercontig_1.20        1507578 1507296 NW_003303739.1
Supercontig_1.62        416848  416616  NW_003303697.1
```

**Comment by jmarshall** [comment]
Is this file perhaps modified from the one in your original report that complained about line 1? Because for me it complains about line 2.

This may be another example where a more specific error message would be helpful (cf https://github.com/samtools/htslib/pull/708#issuecomment-396354102 and the surrounding conversation):

> Parse error reading some.bed at line 2: end (1349821) may not be less than start (1350077)

**Comment by Neato-Nick** [comment]
You're right - I'll edit my original post to say line 2. Just ran it again with the BED I pasted in and got the same error.

Also, thank you! I deleted the regions that ran in reverse (so `end` was less than `start`) and samtools depth parsed the BED correctly. Would it be improper for these tools to allow regions in reverse like this? Or does BED format require that the start has to be before the end?

**Comment by jkbonfield** [comment]
That's a good question on what BED requires.  Sadly it appears to be unspecified: https://genome.ucsc.edu/FAQ/FAQformat.html#format1

I see other software out there that mandates end > beg though so it seems possibly courageous to hope it'll work. :-)

**Comment by jmarshall** [comment]
In the absence of a specy spec, the implementations that could reasonably claim to be a _de facto_ standard would be kent and bedtools. Both of those validate that start ≤ end, so it wouldn't be helpful to be more accepting than that.

**Comment by jmarshall** [comment]
Also clearly the OP has been misled by the spurious _No such file or directory_ message.

This occurs because `bed_read()` fails both for things like `open(2)` that set `errno` informatively and things like semantic data errors that do not. So the other bug here is that all `bed_read()` failure paths need to set `errno` — e.g. for this path either to 0 (to suppress the `perror` message) or to something informative like EINVAL.
